File: concurrency/javascript/counter.js
async function counter() {
    for (let i = 0; i < 10; ++i) {
        console.log(`${i} from ${process.pid}`);
        await sleep(0); // acts like yield
    }
}

function main() {
    counter();
    counter();
}

async function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main();

File: concurrency/python/counter.py
import asyncio
import threading


async def counter():
    for i in range(10):
        print(f"{i} on {threading.current_thread().name}")
        await asyncio.sleep(0)  # acts like yield


async def main():
    await asyncio.gather(counter(), counter())


if __name__ == "__main__":
    asyncio.run(main())
